src/common/dns_utils.cpp

```diff
@@ -266,11 +277,27 @@ DNSResolver::DNSResolver() : m_data(new DNSResolverData())
     ub_ctx_hosts(m_data->m_ub_context, NULL);
   }
 
-  const char * const *ds = ::get_builtin_ds();
-  while (*ds)
+  add_anchors(m_data->m_ub_context);
+
+  if (!DNS_PUBLIC)
   {
-    MINFO("adding trust anchor: " << *ds);
-    ub_ctx_add_ta(m_data->m_ub_context, string_copy(*ds++));
+    // if no DNS_PUBLIC specified, we try a lookup to what we know
+    // should be a valid DNSSEC record, and switch to known good
+    // DNSSEC resolvers if verification fails
+    bool available, valid;
+    static const char *probe_hostname = "updates.moneropulse.org";
+    auto records = get_txt_record(probe_hostname, available, valid);
+    if (!valid)
+    {
+      MINFO("Failed to verify DNSSEC record from " << probe_hostname << ", falling back to well known DNSSEC resolvers");
+      ub_ctx_delete(m_data->m_ub_context);
+      m_data->m_ub_context = ub_ctx_create();
+      add_anchors(m_data->m_ub_context);
+      for (const auto &ip: dns_public_addr)
```

Comment:
Isn't this guaranteed to be empty (`!DNS_PUBLIC` implies that `dns_public_addr` is empty)? This doesn't seem to do anything, and not what the comments indicate. Am I missing something?

Reply:
Good point. I tended to use the global array of well known servers. This implies that it's just the switch to tcp that fixes it, with no need for extra servers... I'll change it to do just that then.

Reply:
Actually, that is not enough. The resolvconf/hosts calls also need to be not made, at least on my setup.

Reply:
I just added the default servers. This seems like the more conservative option.